Add blog entry explaining the solution for day 16 (#79)
scalacenter/scala-advent-of-code: docs/puzzles/day16.md

import Solver from "../../../../website/src/components/Solver.js"

# Day 16: Packet Decoder

by [@tgodzik](https://github.com/tgodzik)

## Puzzle description

https://adventofcode.com/2021/day/16

## Part1: You've got mail!

> It seems that that we can split our problem into two parts. First, we need to

Reviewer: ```suggestion
It seems that we can split our problem into two parts. First, we need to
```

> parse the example into structures that we can later use to calculate out

Reviewer 2: Do you really mean “calculate out” or “calculate our results”?

results.

> Let's start with defining data structures to use:

Reviewer 2: ```suggestion
Let's start with defining the data structures to use:
```

```scala
enum Packet(version: Int, typeId: Int):

>   case Literal(version: Int, exprs: Long) extends Packet(version, 4)

Reviewer 2: Why is it named `exprs` (plural)?

```suggestion
  case Literal(version: Int, value: Long) extends Packet(version, 4)
```

Author: I was reworking the examples and it got lost, thanks!

  case Operator(version: Int, typeId: Int, exprs: List[Packet]) extends Packet(version, typeId)
```

`Packet.Literal` will represent simple literal packets, that contain only a

> value. We are using Long just in case based on the experience with previous

Reviewer 2: just in case of what?

Author: Add more clarification `in case of large integers`

Advent of Code puzzles.

`Packet.Operator` will represent all the other operators that can contain other
packets.

> Now we need to map our input to these structures that we defined.

Reviewer 3: ```suggestion
Now we need to map our input to these structures.
```
"these" + "that" is redundant.

Let's start by mapping out hexadecimal input to list of chars that we can

> analyse easier when checking for packets:

Reviewer 3: ```suggestion
Let's start by mapping the hexadecimal input to a list of chars that we can
analyse easier when checking for packets:
```
"mapping out" means creating an overview. We we're "mapping" instead.

def readOperatorBody(input: BinaryData): (List[Packet], BinaryData) = ???

def decodePacket(packet: BinaryData): (Packet, BinaryData) =
  val (versionBits, rest) = packet.splitAt(3)
  val version = toInt(versionBits)
  val (typeBits, body) = rest.splitAt(3)
  val tpe = toInt(typeBits)

  tpe match
    case 4 =>
      val (value, remaining) = readLiteralBody(body, Nil)
      (Packet.Literal(version, value), remaining)
    case otherTpe =>
      val (values, remaining) = readOperatorBody(body)
      (Packet.Operator(version, otherTpe, values), remaining)
  end match
end decodePacket
```

We use the function `splitAt`, which gives us the ability to split the input
into the part that we need, for example 3 bits for version, and the rest of the
packet data. This way we can read the version and typeId, pattern match on the
latter and use proper logic for reading in each case. We can then create our new
structures. We also defined a helper type `BinaryData`, since we will be using
it throughout the puzzle. What remains is defining `readLiteralBody` and
`readOperatorBody`. You might notice that we return additional `BinaryData` from

> each function. This is because we will be later able to use it to analyze the

Reviewer 2: You sometimes use “analyze” and sometimes “analyse”. I wonder if we should settle on one style.

Reviewer 3: You should definitely settle on one style. As developers we often settle on the US spelling, so "analyze".

Author: Settled on `analyze` :sweat_smile:

output further in a recursive manner, but we'll get back to it.

Let's start with the first undefined function `readLiteralBody`. In the
description we read that the body of the literal consists of segments of 5 bits,

> where the last segment will start with 0 and all the other with 1. The rest 4

Reviewer 3: ```suggestion
where the last segment will start with 0 and all the others with 1. The remaining 4
```

means we will need to apply a recursive approach:

```scala
def readOperatorBody(current: BinaryData): (List[Packet], BinaryData) =
  val (lenId, rest) = current.splitAt(1)

  @tailrec
  def readMaxBits(
      current: BinaryData,
      remaining: Int,
      acc: List[Packet]
  ): (List[Packet], BinaryData) =
    if remaining == 0 then (acc, current)
    else
      val (newExpr, rest) = decodePacket(current)
      readMaxBits(rest, remaining - (current.size - rest.size), acc :+ newExpr)

  @tailrec
  def readMaxPackets(
      current: BinaryData,
      remaining: Int,
      acc: List[Packet]
  ): (List[Packet], BinaryData) =
    if remaining == 0 then (acc, current)
    else
      val (newExpr, rest) = decodePacket(current)
      readMaxPackets(rest, remaining - 1, acc :+ newExpr)

  // read based on length
  if lenId(0) == '0' then
    val (size, packets) = rest.splitAt(15)
    readMaxBits(packets, toInt(size), Nil)
  // read based on number of packages
  else
    val (size, packets) = rest.splitAt(11)
    readMaxPackets(packets, toInt(size), Nil)
  end match
end readOperatorBody
```

In the above function we first check the first bit of the operator body, which

> tell us how we should check the rest of the body.

Reviewer: ```suggestion
tells us how we should check the rest of the body.
```

- if the bit is `0` it means that the next 15 bits can be turned into a number,
  that will define how many of the further bits are the subpackets of the
  operator.

- if the bit is `1` it means that the next 11 bits can be turned into a number,

>   that will defining how many subpackets should belong to the operator.

Reviewer 2: ```suggestion
  that will define how many subpackets should belong to the operator.
```

We defined two helper recursive functions `readMaxBits` and `readMaxPackets`
which will check if the stopping condition (either max bits read or max
packets read) is achieved or read a new packet using recursively the
`decodePacket` function otherwise. At the end they will both return a list of
packets, that we can later use to put into the operator packet, and the
remaining input that we might need to check for more packets.

This should already allow us to create a full structure and what remains is

> adding a function that can sum all the versions up. We can just add that

Reviewer 3: ```suggestion
adding a function that can add up all the versions. We can add that
```

function to the `Packet` enum and sum it all recursively.

```scala
  def versionSum: Int =
    this match
      case Literal(version, _)     => version
      case Operator(version, exprs, _) => version + exprs.map(_.versionSum).sum
```

That's it! We should be able to solve the part 1.

> ### Full solution

Author: Should I maybe just add the full solution at the end?

Reviewer 2: I personally like seeing first the solution of part1

Reviewer: This seems fine, since it is the `Full solution` of part 1.

Author: ok cool!

      readMaxPackets(rest, remaining - 1, acc :+ newExpr)

  lenId match
    // read based on length
    case List('0') =>
      val (size, packets) = rest.splitAt(15)
      readMaxBits(packets, toInt(size), Nil)

    // read based on number of packages
    case _ =>
      val (size, packets) = rest.splitAt(11)
      readMaxPackets(packets, toInt(size), Nil)
  end match
end readOperatorBody

def decodePacket(input: BinaryData): (Packet, BinaryData) =
  val (versionBits, rest) = input.splitAt(3)
  val version = toInt(versionBits)
  val (typeBits, body) = rest.splitAt(3)
  val tpe = toInt(typeBits)

  tpe match
    case 4 =>
      val (value, remaining) = readLiteralBody(body, Nil)
      (Packet.Literal(version, value), remaining)
    case otherTpe =>
      val (values, remaining) = readOperatorBody(body)
      (Packet.Operator(version, otherTpe, values), remaining)
  end match
end decodePacket

def parse(input: String) =
  val number = input.toList.flatMap(hex => hexadecimalMapping(hex).toCharArray)
  val (operator, _) = decodePacket(number)
  operator

def part1(input: String) =
  val packet = parse(input)
  packet.versionSum
```

<Solver puzzle="day16-part1"/>

> ## Part 2: The elven calculus

Reviewer 3: ```suggestion
## Part 2: The Elven calculus
```

  case Literal(version: Int, literalValue: Long) extends Packet(version, 4)
  case GreaterThan(version: Int, lhs: Packet, rhs: Packet) extends Packet(version, 5)
  case LesserThan(version: Int, lhs: Packet, rhs: Packet) extends Packet(version, 6)
  case Equals(version: Int, lhs: Packet, rhs: Packet) extends Packet(version, 7)
```

We will also need to modify the way we create these operators:

So instead of

```scala
      val (values, remaining) = readOperatorBody(body)
      (Packet.Operator(version, otherTpe, values), remaining)
```

we will need to write:

```scala
     val (values, remaining) = readOperatorBody(body)
      otherTpe match
        case 0 => (Packet.Sum(version, values), remaining)
        case 1 => (Packet.Product(version, values), remaining)
        case 2 => (Packet.Minimum(version, values), remaining)
        case 3 => (Packet.Maximum(version, values), remaining)
        case 5 => (Packet.GreaterThan(version, values(0), values(1)), remaining)
        case 6 => (Packet.LesserThan(version, values(0), values(1)), remaining)
        case 7 => (Packet.Equals(version, values(0), values(1)), remaining)
```

> This makes are structure to accurately show the mathematical equation that is

Reviewer: ```suggestion
This makes our structure to accurately show the mathematical equation that is
```

> This makes our structure to accurately show the mathematical equation that is

Reviewer 3: ```suggestion
This makes our structure accurately show the mathematical computation that is
```
It's not an *equation* since there no `=` sign. ;)

constructed from the packets. The last remaining step is to create a function

> that will calculate the equation. We can do it similar to the `versionsSum`

Reviewer 3: ```suggestion
that will calculate the equation. We can do it similarly to the `versionsSum`
```

  val version = toInt(versionBits)
  val (typeBits, body) = rest.splitAt(3)
  val tpe = toInt(typeBits)

  tpe match
    case 4 =>
      val (value, remaining) = readLiteralBody(body, Nil)
      (Packet.Literal(version, value), remaining)
    case otherTpe =>
      val (values, remaining) = readOperatorBody(body)
      otherTpe match
        case 0 => (Packet.Sum(version, values), remaining)
        case 1 => (Packet.Product(version, values), remaining)
        case 2 => (Packet.Minimum(version, values), remaining)
        case 3 => (Packet.Maximum(version, values), remaining)
        case 5 => (Packet.GreaterThan(version, values(0), values(1)), remaining)
        case 6 => (Packet.LesserThan(version, values(0), values(1)), remaining)
        case 7 => (Packet.Equals(version, values(0), values(1)), remaining)
  end match
end decodePacket

def parse(input: String) =
  val number = input.toList.flatMap(hex => hexadecimalMapping(hex).toCharArray)
  val (operator, _) = decodePacket(number)
  operator

def part1(input: String) =
  val packet = parse(input)
  packet.versionSum

def part2(input: String) =
  val packet = parse(input)
  packet.value
end part2

```

> You might have noticed that we had to modify the `versionsSum` function slightly
> to work with our new structure.

Reviewer 3: ```suggestion
You might have noticed that we had to slightly modify the `versionsSum` function
to work with our new structure.
```
